src/bookshelf/src/bksw_sync.erl

```diff
@@ -0,0 +1,227 @@
+          % These paths are stored relative to the base_directory.
+          unsynced_paths,
+          % The base bookshelf data directory.  We initialize this to
+          % bksw_conf:data_store() in init/1,
+          base_directory,
+          % A valid rsync remote URI.  We expect the user to supply
+          % this explicitly as part of the configuration.
+          rsync_remote}).
+
+
+%%
+%% Public API
+%%
+start_link(BaseDir, RemoteURI) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [BaseDir, RemoteURI], []).
+
+new(Path) ->
+    gen_server:cast(?MODULE, {created, Path}).
+
+delete(Path) ->
+    gen_server:cast(?MODULE, {deleted, Path}).
+
+sync() ->
+    gen_server:cast(?MODULE, sync).
+
+full_sync() ->
+    gen_server:cast(?MODULE, full_sync).
+
+status() ->
+    gen_server:call(?MODULE, status).
+
+%%
+%% gen_server callbacks
+%%
+init([BaseDir, RemoteURI]) ->
+    %% Send ourselves a full_sync to ensure that we always do a
+    %% full_sync on startup. Note: this doesn't block init() since it
+    %% is a gen_server:cast underneath.
+    full_sync(),
+    lager:info("starting bksw_rsync for ~p ~p", [BaseDir, RemoteURI]),
+    {ok, #sync_state{unsynced_paths = [],
+                     base_directory = to_binary(BaseDir),
+                     rsync_remote = RemoteURI}}.
+
+handle_cast({created, Path}, State) ->
+    add_to_unsynced(Path, State);
+handle_cast({deleted, Path}, State) ->
+    add_to_unsynced(Path, State);
+handle_cast(full_sync, State) ->
+    lager:info("starting full rsync run"),
+    RsyncCmd = exec:run(mk_rsync_cmd(full, State), [stdout, stderr, sync]),
+    case RsyncCmd of
+        {ok, _Rest} ->
+            lager:info("full rsync run successful."),
+            {noreply, State};
+        {error, Result} ->
+            lager:error("full rsync run failed, retrying in ~p ms", [?MAX_UNFLUSHED_TIME]),
+            log_rsync_error(Result),
+            timer:apply_after(?MAX_UNFLUSHED_TIME, ?MODULE, full_sync, []),
+            {noreply, State}
+    end;
+handle_cast(sync, State = #sync_state{unsynced_paths = []}) ->
+    lager:info("sync message recieved but no data to sync"),
+    {noreply, State};
+handle_cast(sync, State = #sync_state{unsynced_paths = UP}) ->
+    lager:info("starting partial sync on ~p paths", [length(UP)]),
+    {ok, File, Filename} = tempfile(),
+    [file:write(File, iolist_to_binary([P, $\n])) || P <- UP ],
+    file:close(File),
+    Cmd = mk_rsync_cmd(partial, State, Filename),
+    CmdStatus = exec:run(Cmd, [stdout, stderr, sync]),
+    file:delete(Filename),
+    case CmdStatus of
+        {ok, _Rest} ->
+            lager:info("partial rsync run successful."),
+            {noreply, State#sync_state{unsynced_paths = []}};
+        {error, Result} ->
+            lager:error("partial rsync run failed, retrying in ~p ms", [?MAX_UNFLUSHED_TIME]),
+            log_rsync_error(Result),
+            timer:apply_after(?MAX_UNFLUSHED_TIME, ?MODULE, sync, []),
```

Comment:
I was thinking that rather than force a full sync on N failures, force a full sync once the unsync_paths list grows to a certain size.  I think the danger when we are failing a lot of sync attempts in a row is that unsynced_paths keeps growing.